theano/gpuarray/opt.py

```diff
@@ -250,53 +260,69 @@ def apply(self, fgraph):
         # Iterating through inputs of graph
         for i in fgraph.inputs:
             if isinstance(i.type, tensor.TensorType):
-                mapping[i] = GpuFromHost(None)(i)
+                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
             else:
                 mapping[i] = i
-
-        # Iterating through output of all the nodes
-        for n in fgraph.toposort():
-            for o in n.outputs:
-                if isinstance(o.type, tensor.TensorType):
-                    mapping[o] = GpuFromHost(None)(o)
-                else:
-                    mapping[o] = o
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
 
         for node in fgraph.toposort():
 
-            # The Extra condition
-            if any([isinstance(i, GpuArrayVariable) or
-               isinstance(i, GpuArraySharedVariable)
-               for i in node.inputs + node.outputs]):
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
+                continue
 
-                move_to_GPU = False
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = False
+            if any([isinstance(i, GpuArrayVariable) or
+                    isinstance(i, GpuArraySharedVariable)
+                    for i in [mapping[v] for v in node.inputs] +
+                              node.outputs]):
 
-            # Oplifter's condition
-            # Will return a list of OP
-            # If None, means can't be moved.
+                move_to_GPU = True
```

Comment:
We maintain 2 different variables here. I didn't notice it till now. Probably for all the profiling results, we haven't been using this variable at all and rather we check with a different variable(`replace`) in line 296. 
So we have been aggressive all along and I believe all the OPs should have been transferred to the GPU. Am I right ?